velocity-engine-core/src/main/java/org/apache/velocity/runtime/resource/loader/CachingDatabaseObjectsFactory.java

```diff
@@ -0,0 +1,203 @@
+    private int poolsMaxSize;
+    private KeyedObjectPool<String, PreparedStatement> statementsPool;
+    protected Logger log = null;
+
+    private class PreparedStatementFactory  extends BaseKeyedPooledObjectFactory<String, PreparedStatement>
+    {
+        @Override
+        public PreparedStatement create(String sql) throws Exception {
+            checkConnection();
+            return connection.prepareStatement(sql);
+        }
+
+        @Override
+        public PooledObject<PreparedStatement> wrap(PreparedStatement obj) {
+            return new DefaultPooledObject<>(obj);
+        }
+
+        @Override
+        public boolean validateObject(String key, PooledObject<PreparedStatement> obj)
+        {
+            try
+            {
+                return !obj.getObject().isClosed() && obj.getObject().getConnection().isValid(0);
+            }
+            catch (SQLException sqle)
+            {
+                return false;
+            }
+        }
+
+        @Override
+        public void destroyObject(String key, PooledObject<PreparedStatement> p) throws Exception {
+            p.getObject().close();
+        }
+    }
+
+    /**
+     * Initialize the factory with the DataSourceResourceLoader properties
+     * @param dataSource data source
+     */
+    @Override
+    public void init(DataSource dataSource, ExtProperties properties) throws SQLException
+    {
+        this.dataSource = dataSource;
+        this.connection = dataSource.getConnection();
+        this.poolsMaxSize = properties.getInt(STATEMENTS_POOL_MAX_SIZE, STATEMENTS_POOL_MAX_SIZE_DEFAULT);
+        createStatementsPool();
+    }
+
+    private void createStatementsPool()
+    {
+        GenericKeyedObjectPoolConfig<PreparedStatement> poolConfig = new GenericKeyedObjectPoolConfig<>();
+        poolConfig.setMaxTotal(poolsMaxSize);
+        poolConfig.setTestOnBorrow(true);
+        statementsPool = new GenericKeyedObjectPool<>(new PreparedStatementFactory(), poolConfig);
+    }
+
+    @Override
+    public void setLogger(Logger log)
+    {
+        this.log = log;
+    }
+
+    /**
+     * Prepare a statement
+     * @param sql Statement SQL
+     * @return prepared statement
+     */
+    @Override
+    public PreparedStatement prepareStatement(String sql) throws SQLException
+    {
+        try
+        {
+            return statementsPool.borrowObject(sql);
+        }
+        catch (SQLException sqle)
+        {
+            throw sqle;
+        }
+        catch (Exception e)
+        {
+            throw new SQLException("could not prepare statement", e);
+        }
+    }
+
+    protected synchronized void checkConnection() throws SQLException
+    {
+        if (!connection.isValid(0))
+        {
+            reset();
+        }
+    }
+
+    private void reset() throws SQLException
+    {
+        clear();
+        // refresh connection and statements pool
+        connection = dataSource.getConnection();
+        createStatementsPool();
+    }
```

Comment:
Why do you drain the entire pool if just one object is invalid? Wouldn't be better to throw that object away?

Reply:
The private method `refresh()` is only called by `checkConnection()`. If the connection has fallen, all the existing prepared statements cannot be used. By the way we could also just `clear()` the pool, without closing and replacing it, invalid statements will be evicted on validation. But while waiting for the new connection, a borrowed object can be returned on the pool then given back, so we *have* to close the pool - and there's no `repoen()` method, so we have to replace it...

Reply:
I cite the documentation of `GenericKeyedObjectPool`:
> - This method does not destroy or effect in any way instances that are checked out when it is invoked.
> - Invoking this method does not prevent objects being returned to the idle instance pool, even during its execution. Additional instances may be returned while removed items are being destroyed.